content/blog/2019/01/2019-01-07-webhook-firewalls.adoc

```diff
@@ -0,0 +1,115 @@
+tags:
+- jenkins
+- webhooks
+- security
+author: michaelneale
+---
+
+In this post I wanted to show how you can run Jenkins behind a firewall (which could be a corporate firewall, a NAT’ed network like you have at home) but still receive webhooks in real time from GitHub.com. You can generalise this to other services too - such as BitBucket or DockerHub, or anything really that emits webhooks, but the instructions will be for GitHub projects hosted on github.com. 
+
+### What are webhooks
+
+Just a very quick refresher on what webhooks are: Messages (often JSON, but not always) typically posted by HTTP(S) from a server to a client that is listening for events.
+
+image:/images/post-images/2019-01-07-webhook-firewalls/webhooks.png[webhook diagram, role=center]
+
+The events flow left to right, Jenkins sits there happily listing on paths like `/github-webhook/` or `/dockerhub-webhook/` etc for some HTTP request to tell it to wake up and do some work.
+
+GitHub/BitBucket may be reporting a new commit or PR, or DockerHub reporting an upstream image has changed. What all these things have in common is that they push to Jenkins, and expect to be able to push to it (ie that Jenkins is visible to them). This works great when the network is open - say GitHub Enterprise, or Jenkins is listening on the web. 
+
+
+
+
+
+### Not on the web
+
+The trick is when something gets in the middle, say a firewall: 
+
+image:/images/post-images/2019-01-07-webhook-firewalls/firewalls.png[firewall diagram, role=center]
+
+
+(_As is industry standard, all firewalls have to be a wall on fire. Please don’t somehow set bricks on fire in your organisation_)
+
+This is just the same when you fire up Jenkins on your laptop, and want to receive webhooks from github.com (a legitimate thing, perhaps to test out your setup, perhaps to run builds for iOS on a mac, or some corner of a network that is not exposed to the web). Unless your laptop is addressable to the whole web that is (not likely), or your network is configured just right, the webhooks won’t be able to flow. 
+
+This is fine - we can fall back to polling for changes. Except this is terrible. You burn through API quotas, and you don’t get changes in real time, and really no one is happy. 
+
+
+
+
+### Some problems are opportunities
+
+We can both solve this problem, but also, view this is an opportunity. Having things not addressable on the web, or locked down in some default way is a feature, not a bug. You massively reduce your attack surface, and can have defence in depth: 
+
+image:/images/post-images/2019-01-07-webhook-firewalls/exposed.png[exposed on web, role=center]
+
+### A Webhook forwarding service
+
+Enter the memorably* named link:https://smee.io/[Smee]. This is an link:https://github.com/probot/smee[OSS project] provided by GitHub and also helpfully hosted as a service by GitHub. This can capture and forward webhook for you. I’ll try to explain it with a diagram:
```

Comment:
```suggestion
Enter the memorably* named link:https://smee.io/[Smee]. This is an link:https://github.com/probot/smee[OSS project] provided by GitHub and also helpfully hosted as a service by GitHub. This can capture and forward webhooks for you. I’ll try to explain it with a diagram:
```